desmod/simulation.py

```diff
@@ -183,6 +183,7 @@ def simulate(config, top_type, env_type=SimEnvironment, reraise=True,
                 except BaseException as e:
                     env.tracemgr.trace_exception()
                     result['sim.exception'] = repr(e)
+                    top.post_fail()
```

Comment:
Regarding resource teardown, the stated goal for this changeset was to allow components to clean-up in the event of an unhandled exception. I'm using "resource" to refer to anything that would be involved in this clean-up; i.e. in the sense of [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization).

The key issue I'm pointing out is that if a component has resources to finalize/destroy/close/deallocate, then it will ostensibly be doing that either in `post_simulate()` or `get_result()` in the non-exception paths. Since proposed `post_fail_hook()` is called from a `try`/`except` that covers `elaborate`, `run`, `post_simulate`, and `get_result`, that means `post_fail_hook` could end up being called before, during, or after `post_simulate`; or before or during `get_result`. Thus resources may already be deallocated prior to `post_fail_hook()`.

The requirement for deallocation of resources to be idempotent is not particularly unusual, but it does impose some cognitive burden on people writing componets with a `post_fail_hook()`. We should be aware that adding `post_fail_hook` has this cost. And it's docstring should probably highlight the wide variety of states the `top_type` instance might be in when an exception triggers `post_fail_hook`.

----

Regarding the possibility of having a `SimEnvironment.post_fail_hook()`, adding hooks to `SimEnvironment` would be at odds with the existing paradigm for extending `SimEnvironment`'s behavior by subclassing with overriding methods.

As an aside, I chose to provide hooks for `Component` because there is a recursive nature to `elaborate`, `post_simulate`, and `get_result` that would make it some what tricky for users of `Component` to place their super method calls correctly. Using hooks abstracts away that complexity of users needing to know where/when/whether to call the super methods.

`SimEnvironment`, on the other hand, just has a single instance, so it is not such a big deal for a subclass to override its methods. For the problem at hand, overriding `SimEnvironment.run()` would be a reasonable thing for a subclass to do if it wanted to handle simulation-time errors in a special manner.

I wonder if it is sufficient for most models to attach resources that need cleanup to the `SimEnvironment` instance and wrap `SimEnvironment.run()` in a `try/except` block in a `SubEnvironment.run()` method?